mono/utils/mono-error-internals.h

```diff
@@ -43,10 +43,42 @@ struct _MonoErrorBoxed {
 	MonoImage *image;
 };
 
-#define error_init(error) do {	\
-	((MonoErrorInternal*)(error))->error_code = MONO_ERROR_NONE;	\
-	((MonoErrorInternal*)(error))->flags = 0;	\
-} while (0);
+/*
+Historically MonoError initialization was deferred, but always had to occur,
+	even in success paths, as cleanup could be done unconditionally.
+	This was confusing.
+
+ERROR_DECL
+	This is the overwhelmingly common case.
+	Declare and initialize a local variable, typically named "error".
+
+MONO_API_ERROR_INIT
+	This is used for MonoError in/out parameter on a public interface,
+	which must be presumed uninitialized. These are typically
+	marked with MONO_API, MONO_RT_EXTERNAL_ONLY, MONO_PROFILER_API, etc.
+
+error_init_reuse
+	This indicates an error has been cleaned up and will be reused.
+
+error_init_internal
+	Rare cases without a better name.
+	For example, setting up an icall frame, or initializing member data.
+
+new0, calloc, static
+	A zeroed MonoError is valid and initialized.
+	Zeroing an entire MonoError is overkill, unless it is near other
+	bulk zeroing.
+
+All initialization is actually bottlenecked to error_init_internal.
+Different names indicate different scenarios, but the same code.
+*/
+#define ERROR_DECL(x) 			MonoError x; error_init_internal (&x)
+#define error_init_internal(error) 	((error)->init = 0)
```

Comment:
It would be better to be a statement so we don't use it erroneously as we shouldn't expect `error_init_internal` to return a value.

Reply:
I'm not sure but it isn't a big deal either way. What about foo ? error_init_reuse() : 0? Cast to avoid for similar affect? You know, functions are common reuse pattern, and you see people wrapping them in do/while0 to statement-ize, do you?